packages/workbox-google-analytics/src/index.js

```diff
@@ -56,83 +49,122 @@ import replayQueuedRequests from './lib/replay-queued-requests.js';
- *                    expressed as key/value pairs, to be added to replayed
- *                    Google Analytics requests. This can be used to, e.g., set
- *                    a custom dimension indicating that the request was
- *                    replayed.
- * @param {Function=} config.hitFilter
- *                    A function that allows you to modify the hit parameters
- *                    prior to replaying the hit. The function is invoked with
- *                    the original hit's URLSearchParams object as its only
- *                    argument. To abort the hit and prevent it from being
- *                    replayed, throw an error.
- * @memberof module:workbox-google-analytics
+ * Creates the requestWillDequeue callback to be used with the background
+ * sync queue plugin. The callback takes the failed request and adds the
+ * `qt` param based on the current time, as well as applies any other
+ * user-defined hit modifications.
+ * @param {Object} config See workbox.googleAnalytics.initialize.
+ * @return {Function} The requestWillDequeu callback function.
  */
-const initialize = (config) => {
-  config = config || {};
-
-  // Stores whether or not the previous /collect request failed.
-  let previousHitFailed = false;
-
-  self.addEventListener('fetch', (event) => {
-    const url = new URL(event.request.url);
-    const request = event.request;
-
-    if (url.hostname === constants.URL.HOST) {
-      if (url.pathname === constants.URL.COLLECT_PATH) {
-        // If this is a /collect request, then use a network-first strategy,
-        // falling back to queueing the request in IndexedDB.
-
-        // Make a clone of the request before we use it, in case we need
-        // to read the request body later on.
-        const clonedRequest = request.clone();
-
-        event.respondWith(
-          fetch(request).then((response) => {
-            if (previousHitFailed) {
-              replayQueuedRequests(config);
-            }
-            previousHitFailed = false;
-            return response;
-          }, (error) => {
-            logHelper.log('Enqueuing failed request...');
-            previousHitFailed = true;
-            return enqueueRequest(clonedRequest).then(() => Response.error());
-          })
-        );
-      } else if (url.pathname === constants.URL.ANALYTICS_JS_PATH) {
-        // If this is a request for the Google Analytics JavaScript library,
-        // use the network first, falling back to the previously cached copy.
-        event.respondWith(
-          caches.open(constants.CACHE_NAME).then((cache) => {
-            return fetch(request).then((response) => {
-              return cache.put(request, response.clone()).then(() => response);
-            }).catch((error) => {
-              logHelper.error(error);
-              return cache.match(request);
-            });
-          })
-        );
+const createRequestWillDequeueCallback = (config) => {
```

Comment:
When is this method run? the naming doesn't tell me if it's run before the request is removed from the queue to attempt the request or if the request re-attempt has been done and successful and now it should be de-queued.

Reply:
I think it's the `requestWillDequeue` that is most upsetting me.